src/accelerate/utils/transformer_engine.py

```diff
@@ -82,3 +82,41 @@ def has_transformer_engine_layers(model):
         if isinstance(m, (te.LayerNorm, te.Linear, te.TransformerLayer)):
             return True
     return False
+
+
+def contextual_fp8_autocast(fn, fp8_recipe, use_during_eval=False):
+    """
+    Disables FP8 autocast when switching to eval mode. Generally better for accuracy.
+    """
+    from transformer_engine.pytorch import fp8_autocast
+
+    def inner(self, *args, **kwargs):
+        enabled = use_during_eval or self.training
+        with fp8_autocast(enabled=enabled, fp8_recipe=fp8_recipe):
+            return fn(*args, **kwargs)
+
+    return inner
+
+
+def apply_fp8_autowrap(model, fp8_recipe_handler):
```

Comment:
Is it necessary to also provide a method that unwraps the `model` (i.e. restores the `forward` method to the original)? For most use cases, I guess not, as this function basically only ensures that the fp8 autocast context is invoked, but perhaps for serialization or something like that, it could be important to unwrap? Not sure, as it seems that so far there was no such issue.

Reply:
I think `unwrap_model` should catch this, but I'll make a test

Reply:
Made a test to make sure that `unwrap_model` properly gets rid of this.